[scroll-animations] Move some functions from DocumentTimeline to AnimationTimeline

nmoucht opened the pull request:
#### e4d5fc8e99945e2e1e6aed054d54fddb399c1a47
<pre>
[scroll-animations] Move some functions from DocumentTimeline to AnimationTimeline
<a href="https://bugs.webkit.org/show_bug.cgi?id=279466">https://bugs.webkit.org/show_bug.cgi?id=279466</a>
<a href="https://rdar.apple.com/135748611">rdar://135748611</a>

Reviewed by Antoine Quint.

Move some functions on DocumentTimeline to the AnimationTimeline superclass.

* Source/WebCore/animation/AnimationTimeline.cpp:
(WebCore::AnimationTimeline::detachFromDocument):
(WebCore::AnimationTimeline::suspendAnimations):
(WebCore::AnimationTimeline::resumeAnimations):
(WebCore::AnimationTimeline::animationsAreSuspended const):
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::documentWillUpdateAnimationsAndSendEvents):
(WebCore::AnimationTimeline::controller const):
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::detachFromDocument):
(WebCore::DocumentTimeline::suspendAnimations):
(WebCore::DocumentTimeline::resumeAnimations):
(WebCore::DocumentTimeline::animationsAreSuspended const): Deleted.
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::ScrollTimeline):
(WebCore::ScrollTimeline::controller const):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::ViewTimeline):

Canonical link: <a href="https://commits.webkit.org/283574@main">https://commits.webkit.org/283574@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/add33a2e927d0c5fe14fdc5d8ee96cbad31c1b27

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66664 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46039 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19286 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70698 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/17797 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68782 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53838 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/17562 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53416 "Passed tests") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/12004 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69731 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42385 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57689 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34085 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39056 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15080 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16151 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60972 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15421 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/72400 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10621 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14782 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60747 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10653 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57749 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61082 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8741 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2360 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10116 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/41846 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42923 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44106 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42666 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/AnimationTimeline.cpp:
```diff
@@ -82,4 +83,33 @@ std::optional<double> AnimationTimeline::bindingsCurrentTime()
+void AnimationTimeline::detachFromDocument()
+{
+    Ref<AnimationTimeline> protectedThis(*this);
+    if (auto* controller = this->controller())
```
`CheckedPtr`

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/DocumentTimeline.cpp:
```diff
@@ -196,7 +183,7 @@ DocumentTimeline::ShouldUpdateAnimationsAndSendEvents DocumentTimeline::document
 
     if (!wasAnimationResolutionScheduled || animationsAreSuspended() || !shouldRunUpdateAnimationsAndSendEventsIgnoringSuspensionState())
-        return DocumentTimeline::ShouldUpdateAnimationsAndSendEvents::No;
+        return AnimationTimeline::ShouldUpdateAnimationsAndSendEvents::No;
```
This line is right but the method signature should probably be changed from `DocumentTimeline::ShouldUpdateAnimationsAndSendEvents` to `AnimationTimeline::ShouldUpdateAnimationsAndSendEvents`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimeline.h:
```diff
@@ -52,6 +53,17 @@ class AnimationTimeline : public RefCounted<AnimationTimeline>, public CanMakeWe
+    virtual void detachFromDocument();
+
+    enum class ShouldUpdateAnimationsAndSendEvents : bool { No, Yes };
+    virtual ShouldUpdateAnimationsAndSendEvents documentWillUpdateAnimationsAndSendEvents() { return ShouldUpdateAnimationsAndSendEvents::Yes; }
```
I think this should actually default to `No` and `ScrollTimeline` should override to return `Yes` if it has a valid `m_source`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -121,4 +121,11 @@ Ref<CSSValue> ViewTimeline::toCSSValue(const RenderStyle& style) const
+        return AnimationTimeline::ShouldUpdateAnimationsAndSendEvents::Yes;
+    return AnimationTimeline::ShouldUpdateAnimationsAndSendEvents::No;
+}
+
```
This is unnecessary since `ViewTimeline` is a subclass of `ScrollTimeline`.

nmoucht replied:
An important thing to note here is the ViewTimelines will not have their m_source set, just their m_subject. We'll need to compute the m_source by looking up the ancestor chain for the nearest scrollable container (like we do now after https://github.com/graouts/WebKit/commit/691b5a8ea25487df174266f4f39cb55763c452fa). So I think this function should remain separate as ViewTimeline needs to check that m_subject exists (and will eventually check its source, once the code to compute it lands).

graouts commented on Source/WebCore/animation/ScrollTimeline.cpp:
```diff
@@ -121,4 +122,18 @@ Ref<CSSValue> ScrollTimeline::toCSSValue(const RenderStyle&) const
+
+AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ScrollTimeline::documentWillUpdateAnimationsAndSendEvents()
+{
+    if (m_source)
```
I think we should also check that `m_source` is in the tree using `m_source->isConnected()`.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.h:
```diff
@@ -56,6 +56,7 @@ class ViewTimeline final : public ScrollTimeline {
     const CSSNumericValue& startOffset() const { return m_startOffset.get(); }
     const CSSNumericValue& endOffset() const { return m_endOffset.get(); }
     const ViewTimelineInsets& insets() const { return m_insets; }
+    AnimationTimeline::ShouldUpdateAnimationsAndSendEvents documentWillUpdateAnimationsAndSendEvents() override;
```
Remove this.

nmoucht replied:
See above.

graouts commented on Source/WebCore/animation/ScrollTimeline.h:
```diff
@@ -54,6 +55,10 @@ class ScrollTimeline : public AnimationTimeline {
     virtual void dump(TextStream&) const;
     virtual Ref<CSSValue> toCSSValue(const RenderStyle&) const;
 
+    AnimationTimeline::ShouldUpdateAnimationsAndSendEvents documentWillUpdateAnimationsAndSendEvents() override;
+
+    AnimationTimelinesController* controller() const override;
```
These can both be marked `final`.

nmoucht replied:
I think `controller()` should remain separate for now until the code to compute ViewTimeline's source lands.

nmoucht committed:
[scroll-animations] Move some functions from DocumentTimeline to AnimationTimeline
https://bugs.webkit.org/show_bug.cgi?id=279466
rdar://135748611

Reviewed by Antoine Quint.

Move some functions on DocumentTimeline to the AnimationTimeline superclass.

* Source/WebCore/animation/AnimationTimeline.cpp:
(WebCore::AnimationTimeline::detachFromDocument):
(WebCore::AnimationTimeline::suspendAnimations):
(WebCore::AnimationTimeline::resumeAnimations):
(WebCore::AnimationTimeline::animationsAreSuspended const):
* Source/WebCore/animation/AnimationTimeline.h:
(WebCore::AnimationTimeline::documentWillUpdateAnimationsAndSendEvents):
(WebCore::AnimationTimeline::controller const):
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::detachFromDocument):
(WebCore::DocumentTimeline::suspendAnimations):
(WebCore::DocumentTimeline::resumeAnimations):
(WebCore::DocumentTimeline::animationsAreSuspended const): Deleted.
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::ScrollTimeline):
(WebCore::ScrollTimeline::controller const):
* Source/WebCore/animation/ScrollTimeline.h:
* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::ViewTimeline):

Canonical link: https://commits.webkit.org/283574@main
